c10/core/CPUCachingAllocator.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <algorithm>
+#include <deque>
+#include <memory>
+#include <mutex>
+
+#include <c10/core/CPUAllocator.h>
+#include <c10/util/Exception.h>
+#include <c10/util/flat_hash_map.h>
+
+namespace c10 {
+
+class CPUCachingAllocator {
+  /*
+   * What it does:
+   * Caches all the allocations carried out by this allocator.
+   * Cache key is the size of the allocation.
+   * If requested size is found in the cache returns the cached pointer.
+   * What it does not do:
+   * No speculative allocation for any future allocations.
+   */
+  private:
+    std::mutex mutex_;
```

Comment:
I think this needs to be used to guard all accesses to `allocation_map_` and `available_map_`.

Reply:
OK, so I realized that there was one place, `free_cached` where this was not the case, but then it is being used by 1) xctor and 2) allocate_and_cache. The latter one is already protected by lock. The former one is xctor, so that was probably my thought process in the first place. But `free_cached` was exposed before so now I moved it to function.